src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -229,20 +233,103 @@ private static void WriteChunksTo(SourceText sourceText, ObjectWriter writer, in
 
     public static SourceText ReadFrom(ITextFactoryService textService, ObjectReader reader, Encoding? encoding, SourceHashAlgorithm checksumAlgorithm, CancellationToken cancellationToken)
     {
-        using var textReader = ObjectReaderTextReader.Create(reader);
+        using var textReader = CharArrayChunkTextReader.CreateFromObjectReader(reader);
 
         return textService.CreateText(textReader, encoding, checksumAlgorithm, cancellationToken);
     }
 
-    private sealed class ObjectReaderTextReader : TextReaderWithLength
+    public static SourceText CreateSourceText(
+        SyntaxNode node, Encoding? encoding, SourceHashAlgorithm checksumAlgorithm, CancellationToken cancellationToken)
+    {
+        // If this node is small enough to not go into the LOH, we can just fast path directly to creating a SourceText from it.
+        var totalLength = node.FullWidth();
+        if (totalLength <= SourceTextLengthThreshold)
+            return SourceText.From(node.ToFullString(), encoding, checksumAlgorithm);
+
+        // Allocate the space to write the node into.  Explicitly chunked so that nothing goes into the LOH.
+        var chunks = CreateChunks(totalLength);
+
+        // Write the node into that temp space.
+        using var chunkWriter = new CharArrayChunkTextWriter(totalLength, chunks, encoding!, cancellationToken);
+        node.WriteTo(chunkWriter);
+        Contract.ThrowIfTrue(totalLength != chunkWriter.Position);
+
+        // Call into the text service to make us a SourceText from the chunks.  Disposal of this reader will free all
+        // the intermediary chunks we allocated.
+
+        Contract.ThrowIfTrue(chunks.Any(static (c, s) => c.Length != s, CharArrayLength));
+
+        using var chunkReader = new CharArrayChunkTextReader(chunks, totalLength);
+        var result = SourceText.From(chunkReader, totalLength, encoding, checksumAlgorithm);
```

Comment:
>SourceText.From [](http://example.com/codeflow?start=21&length=15)

It feels like there is an inefficiency here. This SourceText.From call takes the reader you have built up over these chunks, detects it's large, and will call LargeText.Decode, which will create chunks over the data. 

I feel like we're jumping through hoops to do this at this level, when maybe it should be at the compiler level.

Reply:
yes.   but we have no efficient paths at all.  We can simultaneously explore a different way of the compiler exposing things (like wrapping an `ImmutableArray<ImmutableArray<char>> chunks` system), but that can happen outdie of this PR.

Note: while any approach *today* involves multiple copies, this approach at least makes one of the copy non LOH and also pooled.

Reply:
Fair enough